### PR TITLE
[Abandoned] Port of Vim's `flatten`

### LINKED ISSUE #12541: `flatten` missing
- `nvim --version`:
```
NVIM v0.4.3
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/tmp/neovim-20191107-85913-1wbgqz6/neovim-0.4.3/build/config -I/tmp/neovim-20191107-85913-1wbgqz6/neovim-0.4.3/src -I/usr/local/include -I/tmp/neovim-20191107-85913-1wbgqz6/neovim-0.4.3/deps-build/include -I/usr/local/opt/gettext/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include -I/tmp/neovim-20191107-85913-1wbgqz6/neovim-0.4.3/build/src/nvim/auto -I/tmp/neovim-20191107-85913-1wbgqz6/neovim-0.4.3/build/include
Compiled by brew@Catalina.local

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.4.3/share/nvim"

Run :checkhealth for more info
```

- `vim -u DEFAULTS` (version: ) behaves differently?
YES

- Operating system/version:
`Darwin 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64`

- Terminal name/version:
xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
:echo flatten([1, [2, [3, 4]], 5])
```

### Actual behaviour
```
E117: Unknown function: flatten
```

### Expected behaviour
```
[1, 2, 3, 4, 5]
```

Not sure if this is a bug or a missing feature (feature request). The documentation for vim (https://vimhelp.org/eval.txt.html) seems to suggest that this function is built in.

### PR DESCRIPTION
**Pull request by cbarrete**
This ports patches 8.2.0935 and 8.2.0937.

Fixes https://github.com/neovim/neovim/issues/12541.

### CODE REVIEW

**Comment by janlazo** [comment]
https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#quickstart
None of the commit descriptions in this PR follow the format used by https://github.com/neovim/neovim/blob/master/scripts/vim-patch.sh so that ported patches can be tracked by Neovim tooling. State your intentions if your patches are full ports of Vim's patches `flatten` or not. It's ok if they're not but always include the Vim commit hash so another developer can finish porting the patches. Update the PR title so that the bot can tag your PR for convenience.

**Comment by cbarrete** [comment]
This is a full port indeed. Is there something else that I should do for this PR to be ready?

**Comment by janlazo** [comment]
Is it difficult to rebase this PR so that the `vim-patch.sh` generates the commit description? I prefer to not add these patches manually in a version.c PR like https://github.com/neovim/neovim/pull/12524 and hunt down related vim patches like `Termdebug` in https://github.com/neovim/neovim/pull/8364.

Tests should pass before this PR can be merged. Sourcehut failures are expected but Travis build should not fail, at least for the first stage. Appveyor builds take a long to finish so but most tests should pass, aside from the flaky async tests like jobs and terminal.

File: src/nvim/eval/funcs.c [hunk 1]
```diff
@@ -2157,6 +2157,43 @@ static void f_expandcmd(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   rettv->vval.v_string = cmdstr;
 }
 
+/*
+ * "flatten(list[, {maxdepth}])" function
+ */
```
Comment (erw7) [inline src/nvim/eval/funcs.c:2162]: There is an error in `clint` as shown in the log below. Please amend this to a comment in `//` format.

https://travis-ci.org/github/neovim/neovim/jobs/703247240#L2373

Comment (cbarrete) [reply src/nvim/eval/funcs.c:2162]: Sure, I just followed the local convention. I'll change that.

**Comment by cbarrete** [comment]
@janlazo I'll just close this PR and start a new one using the script then, it'll likely be easier and cleaner.

**Comment by cbarrete** [comment]
Changing the name of the PR to not interfere with the bot.

**Pull request closed without merging**